### ISSUE TITLE
0.28.0 Upgrade Guide

### ISSUE THREAD

**Issue by CrossEye** [ramda/ramda#3218]
[v0.27.0...v0.28.0](https://github.com/ramda/ramda/compare/v0.27.0...v0.28.0)

(Note that some of these changes were in v0.27.1, which somehow never got its own upgrade guide.)

:new: Added:

- #2852 :new: `on`
- #2881 :new: `splitWhenever`
- #2925 :new: `whereAny`
- #3037 :new: `promap`
- #3047 :new: `partialObject`
- #3107 :new: `modify` and `modifyPath`
- #3125 :new: `unwind`
- #3144 :new: `count`

:x: Removed:

- #2835 
  - :x: `composeK` in favor of `composeWith`
  - :x: `composeP` in favor of `composeWith`
  - :x: `contains` in favor of `includes`
  - :x: `merge` in favor of `mergeRight`, `mergeLeft`, and others
  - :x: `pipeK` in favor of `pipeWith`
  - :x: `pipeP` in favor of `pipeWith`

:bulb: Changes:
- #2594 adds some type safety to `propEq` and `propSatisfies`
- #2837 removes some redundant code from the internal `_isArrayLike`
- #2848 remove unnecessary code from the internal `_equals`
- #2795 fixes an earlier regression in `applySpec`
- #2882 fixes an odd error in `clone`
- #2885 updates `empty` and `isEmpty` to handle typed arrays
- #2962 optimizes the performance of `prop`
- #2986 adds protection against undefined object in `hasIn`
- #3002 extends `filter` to [FantasyLand compatibility](https://github.com/fantasyland/fantasy-land#filterable).
- #3031 updates the the dispatching of `traverse`
- #3048 improves the performance of the internal `_dispatchable`
- #3067 improves `clone`'s handling of Objects with prototypes
- #3081 adds some typechecking to `evolve` and returns non-Objects directly
- #3134 cleans up the code in `adjust`
- #3135 extends `clone` to typed arrays
- #3136 cleans up the code in `groupBy`
- #3177 fixed a potential Denial of Service attack on `trim` (see also #3212)
- #3198 fixes `is` so that `R.is(Object, Object.create(null))` returns `true`
- #3201 adds transducer support to `uniqBy` and `uniqWith`

:star: Miscellaneous

- #2990 added a [Spanish translation](https://github.com/wirecobweb/ramda) of the documentation
- #3192: After a [long process](https://stackoverflow.com/q/69936667) this bug report and suggested fixes was rejected as invalid.  It would still be worth investigating whether there is any possibility that Ramda might be subject to prototype pollution elsewhere.

:clipboard: Documentation fixes: #2637, #2682, #2786, #2871, #2897, #2932, #2939, #2979, #2981, #2989, #2899, #2900, #2914, #2916, #2917, #2919, #2920, #2922, #2927, #2944, #2956, #2958, #2959, #2963, #2968, #2979, #2981, #2989, #2990, #2992, #3005 #3011, #3013, #3014, #3025, #3039, #3053, #3073, #3080, #3083, #3085, #3088, #3109, #3119, #3123, #3139, #3140, #3161, #3170, #3174, #3188, #3199, #3205, #3207, #3208

:heavy_check_mark: Testing improvements:  #2898, #2962, #2984, #3095 #3194, #3205, #3206

:shipit: Build and other internal improvements: #2845, #2928, #2935, #2999, #3030, #3042, #3044, #3046, #3050, #3052, #3054, #3069, #3096

----------

Many thanks to all the contributors supporting this release, especially to those helping with numerous improvements to tests and documentation :tada:, and others improving our build-release process, notably @Andarist, @hitmands, @sebastienfilion, @benbryant0, and @customcommander,  :tada:

We'd also like to call out to all those who submitted issues or PRs that we've chosen *not* to include.  It's often easy to forget that the attempts that fail are often just as important to long-term success as those that succeed.  Thank you for all your great work! :fireworks:

A final call-out to @customcommander, who recently joined the @ramda/core team and has helped push us to getting out this long-delayed release, and ideally to get on a reasonable release cadence.  Thank you!  :heart:

**Comment by SeanCannon** [comment]
Excellent work

**Comment by christophemenager** [comment]
Thanks for the great work !

**Comment by meherranjan** [comment]
Fantastic work team! Cheers!

**Comment by fabianpena90** [comment]
will `merge` still work after the update?

**Comment by customcommander** [comment]
> will `merge` still work after the update?

Nope. As per the release notes `merge` has been removed completely in `0.28`. You can use either `mergeLeft` or `mergeRight` instead

**Comment by customcommander** [comment]
# Replacing `composeP` with `composeWith`

Based from https://github.com/ramda/ramda/issues/3099

Before:

```javascript
import * as R from 'ramda';
import { promises as fs } from 'fs';

await R.composeP(
  JSON.parse,
  R.toString,
  fs.readFile,
)('path/to/file');
```

After: (please note that `composeWith` is no longer a variadic function)

```javascript
await R.composeWith(R.andThen)([
  JSON.parse,
  R.toString,
  fs.readFile,
])('path/to/file');
```

**Comment by sudo-cho** [comment]
Great work! I just noticed the new R.on function is referencing R.contains in its example but it has been removed in favor of includes.

**Comment by jvdneste** [comment]
I noticed that `R.prop(5, { "5": "x" })` returns `"x"` with ramda 0.26 but returns `undefined` with ramda 0.27+
Is this in the upgrade guide? Other than that, awesome work👍

**Comment by johanvandenneste** [comment]
> I noticed that `R.prop(5, { "5": "x" })` returns `"x"` with ramda 0.26 but returns `undefined` with ramda 0.27+ Is this in the upgrade guide? Other than that, awesome work👍

That should be `R.prop(-5, { "-5": "x" })`

**Comment by zydmayday** [comment]
> 

https://github.com/ramda/ramda/commit/13ea6c5532d30e8360a732237bf5b6cd977713a8#diff-7ad2abd5596a4285282263bd8c938729a08476f54ea19bc90d73bf93d60e9ec5

It seems behavior has been changed here.
Treat integer-like prop as index.
